core/trino-main/src/main/java/io/trino/operator/window/matcher/IrRowPatternToProgramRewriter.java

```diff
@@ -0,0 +1,317 @@
+                }
+                else {
+                    startSplitBuilder.setSecond(instructions.size());
+                }
+                int loopingPosition = instructions.size();
+
+                process(pattern);
+                loop(loopingPosition, greedy);
+
+                if (greedy) {
+                    startSplitBuilder.setSecond(instructions.size());
+                }
+                else {
+                    startSplitBuilder.setFirst(instructions.size());
+                }
+                instructions.set(startSplitPosition, startSplitBuilder.build());
+
+                return null;
+            }
+
+            int loopingPosition = instructions.size();
+            for (int i = 0; i < min; i++) {
+                loopingPosition = instructions.size();
+                process(pattern);
+            }
+
+            loop(loopingPosition, greedy);
+
+            return null;
+        }
+
+        private void loop(int loopingPosition, boolean greedy)
+        {
+            Split loopingSplit;
+            if (greedy) {
+                loopingSplit = new Split(loopingPosition, instructions.size() + 1);
+            }
+            else {
+                loopingSplit = new Split(instructions.size() + 1, loopingPosition);
+            }
+            instructions.add(loopingSplit);
+        }
+
+        private Void rangeQuantified(IrRowPattern pattern, boolean greedy, int min, int max)
+        {
+            checkArgument(min <= max, format("invalid range: (%s, %s)", min, max));
+
+            for (int i = 0; i < min; i++) {
+                process(pattern);
+            }
+
+            // handles 0 without adding instructions
+            if (min == max) {
+                return null;
+            }
+
+            List<Integer> splitPositions = new ArrayList<>();
+            List<Split.Builder> splitBuilders = new ArrayList<>();
+
+            for (int i = min; i < max; i++) {
+                splitPositions.add(instructions.size());
+                instructions.add(new SplitDummy());
+                Split.Builder builder = new Split.Builder();
+                if (greedy) {
+                    builder.setFirst(instructions.size());
+                }
+                else {
+                    builder.setSecond(instructions.size());
+                }
+
+                splitBuilders.add(builder);
+                process(pattern);
+            }
+
+            if (greedy) {
+                for (Split.Builder splitBuilder : splitBuilders) {
+                    splitBuilder.setSecond(instructions.size());
+                }
+            }
+            else {
+                for (Split.Builder splitBuilder : splitBuilders) {
+                    splitBuilder.setFirst(instructions.size());
+                }
+            }
+            for (int i = 0; i < splitPositions.size(); i++) {
+                instructions.set(splitPositions.get(i), splitBuilders.get(i).build());
+            }
```

Comment:
This can be simplified:


```java
List<Integer> splitPositions = new ArrayList<>();
List<Integer> splitTargets = new ArrayList<>();

for (int i = min; i < max; i++) {
    splitPositions.add(instructions.size());
    instructions.add(null);
    splitTargets.add(instructions.size());
    process(pattern);
}

for (int i = 0; i < splitPositions.size(); i++) {
    Split split;
    if (greedy) {
        split = new Split(splitTargets.get(i), instructions.size());
    }
    else {
        split = new Split(instructions.size(), splitTargets.get(i));
    }
    instructions.set(splitPositions.get(i), split);
}
```